projects/Mallard/src/helpers/queries.ts

```diff
@@ -0,0 +1,249 @@
+}
+
+type AllowedVariables = { [key: string]: number | string }
+
+export type LocalResolver = <Value, Variables>(
+    query: Query<Value, Variables>,
+    variables: Variables,
+) => Promise<Value>
+
+export type QueryResolver<Value, Variables> = (
+    variables: Variables,
+    resolve: LocalResolver,
+    prevValue: Value,
+) => Promise<Value>
+
+export type QueryResult<Value> =
+    | { loading: true; value: undefined; error: undefined }
+    | { loading: undefined; value: Value; error: undefined }
+    | { loading: undefined; value: undefined; error: unknown }
+
+const formatResult = <Value>(node: QueryNode): QueryResult<Value> => {
+    if (node.value) return { value: node.value as Value } as QueryResult<Value>
+    if (node.error) return { error: node.error } as QueryResult<Value>
+    return { loading: true } as QueryResult<Value>
+}
+
+const EMPTY_STR_SET: Set<string> = new Set()
+
+export class QueryEnvironment {
+    private _graph = new Map<string, QueryNode>()
+
+    /**
+     * Given the specific `query`, call `callback` everytime the value changes.
+     * If the query hasn't been resolved yet, this function will do so.
+     */
+    watch<Value, Variables extends AllowedVariables>(
+        query: Query<Value, Variables>,
+        variables: Variables,
+        listener: (value: QueryResult<Value>) => unknown,
+    ): () => void {
+        const node = this._getNode(query, variables)
+        const callback = () => {
+            listener(formatResult(node))
+        }
+        node.listeners.push(callback)
+        return () => {
+            node.listeners.splice(node.listeners.indexOf(callback), 1)
+            this._releaseNode(node)
+        }
+    }
+
+    peek<Value, Variables extends AllowedVariables>(
+        query: Query<Value, Variables>,
+        variables: Variables,
+    ): QueryResult<Value> {
+        const key = this._getKey(query, variables)
+        const node = this._graph.get(key)
+        if (node === undefined) return { loading: true } as QueryResult<Value>
+        return formatResult(node)
+    }
+
+    invalidate<Value, Variables extends AllowedVariables>(
+        query: Query<Value, Variables>,
+        variables: Variables,
+    ): void {
+        const key = this._getKey(query, variables)
+        const node = this._graph.get(key)
+        if (node === undefined) return
+        this._refresh(node, EMPTY_STR_SET)
+    }
+
+    private _getKey<Value, Variables>(
+        query: Query<Value, Variables>,
+        variables: Variables,
+    ) {
+        return query.id + ':' + JSON.stringify(variables)
+    }
+
+    private _getNode<Value, Variables>(
+        query: Query<Value, Variables>,
+        variables: Variables,
+    ): QueryNode {
+        const key = this._getKey(query, variables)
+        let node = this._graph.get(key)
+        if (node !== undefined) return node
+
+        node = {
+            key,
+            query: (query as unknown) as Query<unknown, unknown>,
```

Comment:
So that we can assign it to the `query` field of the node. The `query` field cannot be strongly typed, because it's then added to the global graph. There is no way to strongly type nodes in the graph since it mixes various queries, so it's a classic case of code where we have to enforce typing at the API level, but we must workaround the typesystem internally. In other languages, workarounds would be for example to use dynamic interfaces, abstract classes, or whatnot.

Reply:
It was more about the return type of this function rather than what lives in the global graph although I can see that this is still managing to be properly typed, even with the unknown return value so looks good: https://github.com/guardian/editions/pull/854/files/855e53d20479c79fe429d29f627acd91412023b5#diff-ff6584e5f0e644cfd7e6bcabb22b845aR53